### DISCUSSION TITLE
Support for Type Families

### DISCUSSION THREAD

**Discussion by newhoggy** [haskell/haskell-language-server#1637] [Ideas]
At the moment, type families look like a regular type constructor.  For example, in the following tooltip in HLS:

<img width="542" alt="Screen Shot 2021-03-30 at 3 41 55 pm" src="https://user-images.githubusercontent.com/63014/112935761-2949d000-9170-11eb-9fee-32bf41ee1a20.png">

It is non-obvious that `AuxiliaryData` is a type family.

If there was some visual cue that `AuxiliaryData` is a type family, for example by being rendered in a different colour, then the type would be much easier to understand.

Even better would be if the type family was expanded.  For example while type above is shown to be `Maybe (AuxiliaryData (ShelleyLedgerEra era))`, the type below could be expanded to say what the type is after type family application, which is to 
`Maybe (Metadata (ShelleyEra StandardCrypto))`

Even more better would be if I could see each step of type family application:

```
Maybe (AuxiliaryData (ShelleyLedgerEra era))
Maybe (AuxiliaryData (ShelleyEra StandardCrypto))
Maybe (AuxiliaryData StandardShelley)
Maybe (Metadata (ShelleyEra StandardCrypto))
```

**Comment by googleson78** [comment]
Actually there is a case where this does currently work - when asking for the type of a term in a function body - it will display the declared (type family) type, and the "current" one, at the point at which you asked for it.
![2021-03-30-111949_2560x1440_scrot](https://user-images.githubusercontent.com/8667384/112957175-d9f0a900-9149-11eb-967c-a3f472dc6294.png)

**Comment by michaelpj** [comment]
I think we'd need semantic highlighting to make it look different. That needs some work in the `lsp` library and then we'd have to use it in HLS. So not totally trivial but something we should take advantage of at some point.

**Comment by soulomoon** [comment]
We have semantic highlighting now. should be easy to differ them now. If they are in the source code.
<img width="467" alt="image" src="https://github.com/haskell/haskell-language-server/assets/14073857/106b3df4-0c72-4f4a-ac48-bc0ba3512991">
